Sources/ArcGISToolkit/Components/Offline/OfflineMapViewModel.swift

```diff
@@ -112,25 +120,51 @@ class OfflineMapViewModel: ObservableObject {
         
         // Note: We don't reset the mode once it is determined.
         
-        // First load preplanned map models.
-        if mode == .undetermined || mode == .preplanned {
+        switch mode {
+        case .preplanned:
             await loadPreplannedMapModels()
-            if mode == .undetermined, hasAnyPreplannedMapAreas {
-                // If there are any preplanned map areas at all
-                // and the mode is undetermined, then set mode to preplanned.
+        case .onDemand:
+            await loadOnDemandMapModels()
+        case .noInternetAvailable, .ambiguous:
+            await determineMode()
+        }
+        
+        /// Attempts to determine the mode.
+        /// - Precondition: `mode == .ambiguous || mode == .noInternetAvailable`
+        func determineMode() async {
+            precondition(mode == .ambiguous || mode == .noInternetAvailable)
+            
+            // Start by trying to load preplanned map models.
+            await loadPreplannedMapModels()
+            
+            if hasAnyPreplannedMapAreas {
+                // If there are any preplanned map areas, set mode to preplanned.
                 mode = .preplanned
+                return
             }
-        }
-                
-        // Load on-demand map models if mode is on-demand or still
-        // undetermined.
-        if mode == .undetermined || mode == .onDemand {
+            
+            // Try to load on-demand map models.
             await loadOnDemandMapModels()
-            // If there are any on-demand areas at all, and the mode is
-            // undetermined, then set mode to on-demand.
-            if mode == .undetermined, !onDemandMapModels.isEmpty {
+            
+            // Try to load on-demand map models.
+            // If there are on-demand areas on device, set mode to on-demand.
+            if !onDemandMapModels.isEmpty {
                 mode = .onDemand
+                return
+            }
+            
+            if !hasAnyPreplannedMapAreas, isShowingOnlyOfflineModels {
```

Comment:
I guess the `!hasAnyPreplannedMapAreas` isn't necessary since it's checked above.

```suggestion
            if isShowingOnlyOfflineModels {
```